ystia/yorc
# Feature/gh 205 hp consumable resources

# Pull Request description
Allow to specify host generic resources capabilities for hosts pool compute.
This resources are consumable by default but can be defined as **no consumable**.

## Description of the change
### What I did

### Tosca stuff
Extend **tosca.capabilities.Container** with **yorc.capabilities.hostspool.Container** to add a **resources** property
the **resources** property is a list of **yorc.datatypes.hostspool.GenericResource**
A **resource** is defined with **name**, **ids**, **number** and **no_consumable** properties.
- **ids** is a list of comma-separated required generic resource ID's for each node instance.
- **number** is the number of generic resource required.
Either **ids** or **number** must be filled to define the resource need.

- **no_consumable** allows to specify if a resource can be only used by a single compute or shared by  others.By default, the generic resource is consumable


### CLI stuff
Display generic resources for each host allocation

###  Deployment Install
- Add host filters for generic resources
- allocate generic resources to host
- Update host resources labels after allocation
- Add instance attribute for generic resource name

###  Deployment uninstall
- Update host resources labels after release

###  Hosts pool config changes
- Recalcul host generic resources with its allocations


### How to verify it
Doc: https://yorc.readthedocs.io/en/feature-gh-205-hp-consumable-resources/infrastructures.html
CI feature: https://github.com/ystia/yorc/blob/feature/godog-ci/testdata/ci/features/GenericResource.feature

Apply the hosts pool config: [hp_conf.yaml.zip](https://github.com/ystia/yorc/files/4191418/hp_conf.yaml.zip)

The pool contains 2 hosts:
- one with:: host.resource.gpu: "gpu2"
- one with; host.resource.gpu: "gpu0,gpu1"

Deploy the following app: [deployment.zip](https://github.com/ystia/yorc/files/4194868/deployment.zip)
with 

`yorc d deploy deployment.zip --id test`

The App contains 2 computes:
- Compute requires 1 **gpu** with a specified id  "gpu2"
- Compute2 requires 2 **gpus** without specified ids.

Check the computes are allocated to the right host according to its **gpu** requirement.
Check **gpu** attribute is correctly set
Check the **host.resource.gpu l**abel is correctly updated for each host

Undeploy the app and check the gpus are correctly released with

`yorc d undeploy -p test`

The Yorc related documentation is very close to this example.

### Description for the changelog
* Add a consumable resources concept to the Hosts Pool ([GH-205](https://github.com/ystia/yorc/issues/205))
## Applicable Issues
fixes #205
